include/boost/math/tools/polynomial.hpp

```diff
@@ -387,30 +387,37 @@ class polynomial :
    }
 
    template <class U>
-   polynomial& operator %=(const U& /*value*/)
+   polynomial& operator %=(const U& value)
    {
-       // We can always divide by a scalar, so there is no remainder:
-       *this = zero_element(std::multiplies<polynomial>());
+       // In the case that T is integral, this preserves the semantics
+       // p == r*(p/r) + (p % r), for polynomial<T> p and U r.
+       if (std::numeric_limits<T>::is_integer)
```

Comment:
Yes, I was right: GCC 6.1, GCC 5.3, and Clang 703.0.31 all complain vociferously. To separate them, it's necessary to add a dummy condition to the enable_if that depends on `U`, e.g. go back to `BOOST_DEDUCED_TYPENAME enable_if_c<boost::is_convertible<U,T>::value &&
std::numeric_limits<T>::is_integer, polynomial >::type& operator %=(const U& value)`.

I don't think this mess is worth it:
- the modulus operator on a polynomial doesn't seem performance-critical
- a conditional testing a compile-time boolean should only emit the chosen branch anyway